Feat/io2021 c 17 employee supervisor insert form

Pinioo opened the pull request:
Goals:
- Providing user insert mechanism [front]
- Providing sufficient interface for user insertion [back]

Pinioo committed:
Simple user insert form created
- Form
- Route to form via /add
- User and PersonalData changed so ids are nullable (for endpoint compatibility)

Pinioo committed:
Very ugly, working user insertion

Pinioo committed:
User insertion form with all necessary fields
- UserInsertionData type containing all user details
- userService addUser now takes UserInsertionData as a parameter

Pinioo committed:
Merge branch 'develop' into feat/IO2021C-17-Employee-Supervisor-Insert-Form

# Conflicts:
#	react-app/package-lock.json
#	react-app/src/components/Routes.tsx
#	react-app/src/services/userService.ts
#	spring-app/src/main/java/com/agh/hr/controllers/UserController.java
#	spring-app/src/main/java/com/agh/hr/persistence/dto/Converters.java
#	spring-app/src/main/java/com/agh/hr/persistence/dto/UserDTO.java
#	spring-app/src/main/java/com/agh/hr/persistence/service/RoleService.java

Pinioo committed:
User insert form completed and refactor
Front:
- UserFormDate added (integrating form hooks with KeyboardDatePicker)
- UserInsertForm is now UserForm like
- EmployeesList add button now routes to insertion form
- Added route '/employees/insert' to insertion form with layout
- [eslint] 'react-hooks/exhaustive-deps' rule set to 'warn'

Back:
- UserInsertionDTO added with all info needed to add new user
- insertUser and saveUser now works with UserInsertionDTO
- saveUser sets passwordHash
- saveUser sets default values for enabled, authorities and permissions
- PasswordEncoder bean moved to separate file (due to cyclic dependencies in SecurityConfig)
- [Tests] tests adjusted to work with UserInsertionDTO

ChrisMisan commented on spring-app/src/main/java/com/agh/hr/persistence/service/UserService.java:
```diff
@@ -49,15 +51,20 @@ public UserService(UserRepository userRepository,RoleService roleService, Conver
-    public Optional<UserDTO> saveUser(UserDTO userDTO) {
+    public Optional<UserDTO> saveUser(UserInsertionDTO userDTO) {
         User user=converters.DTOToUser(userDTO);
+        System.out.println(user.toString());
```
delete

Pinioo replied:
done

ChrisMisan commented on spring-app/src/main/java/com/agh/hr/persistence/service/UserService.java:
```diff
@@ -49,15 +51,20 @@ public UserService(UserRepository userRepository,RoleService roleService, Conver
                 return Optional.empty();
 
         user.setId(0L);
+        user.setPasswordHash(passwordEncoder.encode(userDTO.getPassword()));
+        user.setEnabled(true);
+        user.setAuthorities(Collections.singleton(roleService.employeeRole()));
+        user.setPermissions(Permission.builder().add(false).read(null).write(null).build());
```
Extract this to corresponding converter.

Pinioo replied:
done

ChrisMisan reviewed (changes requested):
Reviewed backend part only.

hardkov commented on react-app/src/components/Profile/UserForm/UserFormDate.tsx:
```diff
@@ -0,0 +1,54 @@
+  disabled: boolean;
+}
+
+const UserFormField: React.FC<Props> = ({ setValue, fieldName, textName, disabled }: Props) => {
```
Change component name to match filename.

hardkov commented on react-app/src/components/Profile/UserForm/UserFormDate.tsx:
```diff
@@ -0,0 +1,54 @@
+  disabled: boolean;
+}
+
+const UserFormField: React.FC<Props> = ({ setValue, fieldName, textName, disabled }: Props) => {
```
Change component name to match file name.

Pinioo replied:
done

Pinioo committed:
Review comments applied
- Printing from saveUser removed
- Default user properties moved to converter
- Names conflict in UserFormDate resolved
- Password input in insert form is now of type password

Pinioo commented on react-app/src/components/UserInsertForm.tsx:
```diff
@@ -0,0 +1,67 @@
+          <UserFormField register={register} fieldName="personalData.address" textName="Address" disabled={disabled} />
+          <UserFormDate setValue={setValue} fieldName="personalData.birthdate" textName="Birthdate" disabled={disabled} />
+          <UserFormField register={register} fieldName="username" textName="Username" disabled={disabled} />
+          <UserFormField register={register} fieldName="password" textName="Password" disabled={disabled} />
```
Password input should be of "password" type

Pinioo committed:
Tests repaired

hardkov reviewed (approved):
Looks good.
